### ISSUE TITLE
Proposal: bwapi-static library

### ISSUE THREAD

**Issue by ratiotile** [bwapi/bwapi#626]
Extract static, hardcoded information that does not depend on hooks into BW should be into a separate library module. This module will expose both C++ and C interfaces. It will be consumed by the rest of the BWAPI library through its C++ interface. The C ABI interface will allow linkage by code produced from different C++ compilers, as well as other languages.

One problem that this solves is the implementation of a bot (with incompatible c++ runtime) directly using the shared memory client interface, which would have required hardcoding the same type information as already exists in BWAPI-LIB. With static information extracted to a dll, both BWAPI and the client bot can share the same static information.

This proposal would not change the existing BWAPI interface functions. It would add a new header file and library which would be optional for bot creators to use.

**Comment by heinermann** [comment]
It could also detach BWAPI for unit testing purposes, so I'd make that another requirement.

I simply don't have the time to work on it these days though.

Maybe I should use vacation days to work on BWAPI?

**Comment by heinermann** [comment]
I'd want to do design documents regarding major changes from here on out though, as BWAPI slowly becomes a more professional project.

**Comment by ratiotile** [comment]
I'm willing to work on this, but I want to reach consensus before starting. What sort of design document do you have in mind? Putting the api specs into this issue tracker is a reasonable approach. I don't think it's complicated enough or has enough people involved to create a separate github repo for the design document.

**Comment by heinermann** [comment]
We can use http://www.starcraftai.com/ or the Github wiki to wiki-edit the design itself while throwing ideas around in this issue.

Oh so from my understanding you want to make the majority of BWAPILIB a DLL so that static data(UnitType, etc) can be accessed through it? I thought you were considering converting the BWAPI interface itself into C.

**Comment by kvakvs** [comment]
BWAPI static library can be built via CMake is now merged into develop.
It is not perfect though, and not 100% GCC/Clang friendly yet but we're getting there.

The original idea of "extracting static hardcoded information" is sort of preserved, you get the static library which makes linking client possible. But it also compiles everything else, like microsoft assembly instructions to patch the Broodwar, many of them are not recognized by GCC/Clang, and are not important for client anyway. I will eventually work more on these CMake files.

**Comment by nturley** [comment]
I worked around this problem by dumping all of the static data into a giant json file (using reflection on the java bindings). That way I can tinker with build order simulations offline in any language on any platform. But having the same API for offline and online processing would obviously be better.

**Comment by lionax** [comment]
How is the current state of this feature?

Background is, I have created [docker-starcraft](https://github.com/docker-starcraft) and [docker-botcraft](https://github.com/docker-botcraft) as a way to compile and run StarCraft bots on any docker compliant operating system. The only problem by now is, that though I am able to cross-compile for example Steamhammer (with some fixes) I'm not able to link against BWAPI 4.1.2.

How far is the support for linking g++ compiled code against BWAPI 4.2.0? What is current work I could help with?

I tried to rewrite some code to make cross-compile possible and failed at the VS assembly code sections. The proposed idea from this issue is way better I suppose.
